iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp

```diff
@@ -0,0 +1,211 @@
+    //                 at compile time.
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays, hicpp-explicit-conversions)
+    static expected<Child, SemanticStringError> create(const char (&value)[N]) noexcept;
+
+    /// @brief Creates a new SemanticString from the provided string.
+    ///         If the value contains invalid characters or invalid content
+    ///         the expected returns an error describing the cause.
+    /// @param[in] value the value of the SemanticString
+    /// @return expected either containing the new SemanticString or an error
+    template <uint64_t N>
+    static expected<Child, SemanticStringError> create(const string<N>& value) noexcept;
+
+    /// @brief Returns the number of characters.
+    /// @return number of characters
+    constexpr uint64_t size() const noexcept;
+
+    /// @brief Returns the capacity of the string.
+    /// @return the maximum amount of characters which can be stored.
+    static constexpr uint64_t capacity() noexcept;
+
+    /// @brief Returns a const reference to the underlying string. It is const
+    ///         and shall not be modified to guarantee the contract that a
+    ///         SemanticString contains always a valid value.
+    /// @return string reference containing the actual value.
+    constexpr const string<Capacity>& as_string() const noexcept;
+
+    /// @brief Appends another string to the SemanticString. If the value contains
+    ///        invalid characters or the result would end up in invalid content
+    ///        it fails.
+    /// @param[in] value the value which should be added
+    /// @return on failure the error inside the expected describes the failure
+    template <typename T>
+    expected<SemanticStringError> append(const T& value) noexcept;
+
+    /// @brief Inserts another string into the SemanticString. If the value contains
+    ///        invalid characters or the result would end up in invalid content
+    ///        it fails.
+    /// @param[in] pos the position where the string should be inserted
+    /// @param[in] str the value which should be added
+    /// @param[in] count how many characters of str shall be inserted
+    /// @return on failure the error inside the expected describes the failure
+    template <typename T>
+    expected<SemanticStringError> insert(const uint64_t pos, const T& str, const uint64_t count) noexcept;
+
+    /// @brief checks if another SemanticString is equal to this string
+    /// @param [in] rhs the other SemanticString
+    /// @return true if the contents are equal, otherwise false
+    bool operator==(const SemanticString& rhs) const noexcept;
+
+    /// @brief checks if another string or char array is equal to this string
+    /// @param [in] rhs the other string
+    /// @return true if the contents are equal, otherwise false
+    template <typename T>
+    IsStringOrCharArray<T, bool> operator==(const T& rhs) const noexcept;
```

Comment:
It would be something like
```cpp
if (username1 == filename2) {
   // do something with username2
}
```

Reply:
I will think and sleep about it.

Reply:
I was just a food for thought with the idea that it is always easier to add than to take away. No strong opinion in either direction.

Reply:
The more I think about it the more I like it when the comparison works out of the box. I see all over the place use cases for comparing different semantic strings. And the beauty of the semantic string is that it cannot be easily mixed up with the actual functions which do things ...
Of course, the problem you drafted is a problem but if feels weird since the user mixes up variable names and this can happen also at any time any where and I would not try to improve the safety for this ... this is a task without end and not too much gain.